test/functional/rpc_rawtransaction.py

```diff
@@ -74,12 +75,11 @@ def setup_network(self):
         self.connect_nodes(0, 2)
 
     def sync_peers(self):
-        self.sync_mempools()
+        self.sync_mempools(self.nodes[0:3])  # nodes 0 to 3 have a mempool
```

Comment:
The -blocksonly nodes allow us to add missing test coverage for rpc getrawtransaction error cases like in my https://github.com/bitcoin/bitcoin/pull/22437#discussion_r696812484 above. We could maybe pass non-existing txids instead, but it seems better to test actual -blocksonly behavior.

I have a branch to add that coverage and some other improvements, and rather than increase the size of this pull and invalidate the existing review, I can propose it as a follow-up.

Reply:
> it seems better to test actual -blocksonly behavior.

blocksonly is a network option, so it seems confusing to test it's behaviour inside a raw transaction test.

Reply:
My thought was that -blocksonly effectively disables mempool functionality, which for e2p testing relates to how getrawtransaction behaves when a node operator running -blocksonly calls getrawtransaction, which is valid with either txindex and/or passing blockhash if the tx is in a block, so it seems good to have explicit coverage that serves as a regression test and sanity check. Passing a wrong txid could be done where relevant to verify the same outcome.

Reply:
> -blocksonly effectively disables mempool functionality

The behaviour of the mempool module isn't affected by blocksonly at all. (All txs that are submitted to the mempool are added/rejected in the same way regardless of blocksonly.) blocksonly is purely a net-processing option to disable *incoming* tx relay from network peers. It doesn't affect (in)validity of txs submitted to the mempool. For example, when the wallet or rpc submits a tx, it will still be added to the mempool.

As part of this test, if the goal is to get a fully functioning, but empty mempool, it might be easier to run the tests that you want to run at the beginning with an empty mempool. (Or by restarting a node with `-nopersistmempool` to empty the mempool). Or as mentioned previously by picking a txid that is not currently in the mempool.


If there was an option to completely disable the mempool (i.e. never initialize the class), then it indeed would make a lot of sense to check for that code path.

In testing it is important to be able to draw units/modules and assume that the modules don't interact with each other, other than over their defined interface. This allows to test units/modules separate from each other. If they had to be tested in combination it would lead to path explosion in any real software project, thus making it harder to test meaningful.

Moreover, if the behaviour of one module is required for testing another module, it will make it harder to change the behaviour of that module and only touching the tests of that module.

Or put another way: If `-blockonly` was relevant to test here, why is `-minrelaytxfee` (or any other setting) not relevant to test here?

Reply:
I agree from a unit test point of view, but for end-to-end testing these interactions can be good to test. For example, I think I've found a parameter interaction that doesn't work as documented (not sure yet, need to look further). That seems like good coverage to have if someone proposes it, it is reviewed, and ready to use.  The asmap/addrman interaction is another example.  What to do here?  Lots of time invested so far.  Sure, I could spend even more time to rework this without the coverage and hope for everyone to review it again.  Or just drop the work starting with "Test...with -blocksonly" and maybe pick up the rest later in a follow-up.

Reply:
> The asmap/addrman interaction is another example

I think asmap is part of the addrman module, so testing those together in a test makes sense.

Reply:
Thank u